lib/Sema/BuilderTransform.cpp

```diff
@@ -1853,33 +1853,66 @@ std::vector<ReturnStmt *> TypeChecker::findReturnStatements(AnyFunctionRef fn) {
 
 bool TypeChecker::typeSupportsBuilderOp(
     Type builderType, DeclContext *dc, Identifier fnName,
-    ArrayRef<Identifier> argLabels, SmallVectorImpl<ValueDecl *> *allResults) {
+    ArrayRef<Identifier> argLabels,
+    SmallVectorImpl<BuilderOpMismatch> *nearMisses) {
+
+  auto recordNearMiss = [&](ValueDecl *vd, BuilderOpMismatchKind kind) {
+    if (nearMisses)
+      nearMisses->emplace_back(vd, kind);
+  };
+
   bool foundMatch = false;
   SmallVector<ValueDecl *, 4> foundDecls;
   dc->lookupQualified(
       builderType, DeclNameRef(fnName),
       NL_QualifiedDefault | NL_ProtocolMembers, foundDecls);
   for (auto decl : foundDecls) {
-    if (auto func = dyn_cast<FuncDecl>(decl)) {
-      // Function must be static.
-      if (!func->isStatic())
-        continue;
+    if (isa<EnumElementDecl>(decl)) {
+      recordNearMiss(decl, BuilderOpMismatchKind::EnumCase);
+      continue;
+    } else if (!isa<FuncDecl>(decl)) {
+      recordNearMiss(decl, BuilderOpMismatchKind::Other);
+      continue;
+    }
 
-      // Function must have the right argument labels, if provided.
-      if (!argLabels.empty()) {
-        auto funcLabels = func->getName().getArgumentNames();
-        if (argLabels.size() > funcLabels.size() ||
-            funcLabels.slice(0, argLabels.size()) != argLabels)
-          continue;
+    auto *func = cast<FuncDecl>(decl);
+    // Function must be static.
+    if (!func->isStatic()) {
+      recordNearMiss(func, BuilderOpMismatchKind::InstanceMethod);
+      continue;
+    }
+
+    // Function must have the right argument labels, if provided.
+    if (!argLabels.empty()) {
+      auto funcLabels = func->getName().getArgumentNames();
+      if (argLabels.size() > funcLabels.size() ||
+          funcLabels.slice(0, argLabels.size()) != argLabels) {
+        recordNearMiss(func, BuilderOpMismatchKind::ArgumentLabelMismatch);
+        continue;
       }
+    }
 
-      foundMatch = true;
-      break;
+    // After the labels provided by the builder transform, any additonal labeled
+    // parameters must have a default value (or else the function will be
+    // uncallable by the builder transformation).
+    auto params = func->getParameters()->getArray();
+    auto isLabeledUndefaulted = [&](ParamDecl *param) -> bool {
+      return !param->hasDefaultExpr() && !param->getArgumentName().empty();
```

Comment:
Please use `isDefaultArgument` rather than `hasDefaultExpr`

Reply:
(Alternatively, using `matchCallArguments` will do the same thing as a real call)

Reply:
See my other comment re: `matchCallArguments`.